### PR TITLE
CLN: simplify core.algorithms

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -652,17 +651,13 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
...
         dtype = original.dtype
     else:
         values, dtype, _ = _ensure_data(values)
 
-        if (
-            is_datetime64_any_dtype(original)
-            or is_timedelta64_dtype(original)
-            or is_period_dtype(original)
-        ):
+        if original.dtype.kind in ["m", "M"]:
```
Comment (WillAyd) [inline pandas/core/algorithms.py]: Perhaps a silly question but wouldn't the datetimelike arrays pass through the `is_extension_array_dtype` branch here?

Comment (jbrockmendel) [reply pandas/core/algorithms.py]: the period and dt64tz go through that EA branch above, td64 and dt64-naive go through here

Comment (jreback) [reply pandas/core/algorithms.py]: can you add a comment to that effect. (can also merge this and do that later)

Comment (jbrockmendel) [reply pandas/core/algorithms.py]: will comment in next pass

**Comment by jreback** [comment]
thanks

**Comment by jbrockmendel** [comment]
I'll take a look, thanks

**Pull request merged**